[match] Better Windows (and Linux) support

janpio opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
`match`/`sync_code_signing` doesn't work on Windows (and Linux) right now.

### Description
This PR improves the support by fixing quite a few bugs and moving Mac only code behind `if`s.

- Fix the `git` commands to clone the match git repo
- Fix the reading of certificate information
- Use `openssl` as a fallback when decoding provisioning profiles if not on Mac (and `security` does not exist)
- Skip the installation of certificate, key and provisioning profile into the Keychain if not on Mac
- Add a new `output_path` option where certificate, key and provisionig profile are saved to (which is necessary on non Mac machines, as there is no import into the non existing Keychain where one could export those manually)
- Fix a general crash when running match (with readonly) for types that are not in storage yet (closes #13965)
- Fix encryption/openssl test so this can actually fail on Windows where this currently is broken: Change encryption/openssl test to work on real provisioningprofile (binary data) + properly read files so comparison can work
- Fix encryption (on Windows)
- Skip sigh certificate verification on non-Mac

janpio committed:
Fix git clone command to use shellescaped params

janpio committed:
Helper.with_env_values

janpio committed:
replace ENV var per command with Helper.with_env_values

janpio committed:
adapt git storage tests to new shellescaped and un-GIT_TERMINAL_PROMPTed commands

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

janpio committed:
Use Helper.with_env_values for GIT_TERMINAL_PROMPT as well, fix several commands to use shellescape for their params

janpio committed:
Fix: Read certificate as binary data
Better error output

janpio committed:
use openssl to decode provisioning profile on non-Mac

janpio committed:
Skip certificate installation on non-Mac

janpio committed:
Skip installation of provisioning profile on non-Mac

janpio committed:
rubocop

janpio committed:
new option :export_path to export decrypted files to a path

janpio committed:
better location for security fallback code + parsing can now be the same on Windows and Mac again

janpio committed:
Fix crash when running match (with readonly) for types that are not in storage yet

closes #13965

janpio committed:
Fix encryption/openssl test

So this can actually fail on Windows where this currently is broken:
Change encryption/openssl test to work on real provisioningprofile (binary data)
Properly read files so comparison can work

janpio committed:
Fix encryption (was broken on Windows)

janpio committed:
skip sigh certificate verification on non-Mac

janpio committed:
remove TODOs

snatchev commented on fastlane_core/lib/fastlane_core/helper.rb:
```diff
@@ -299,6 +299,22 @@ def self.zip_directory(path, output_path, contents_only: false, overwrite: false
+    # Executes the provided block after adjusting the ENV to have the
+    # provided keys and values set as defined in hash. After the block
+    # completes, restores the ENV to its previous state.
+    def self.with_env_values(hash, &block)
```
Nice!

janpio replied:
Oh, I fully stole that from the screengrab `spec_helper`:
https://github.com/fastlane/fastlane/blob/89da896d167f8c7ec33c05601ffe37a182e788d1/screengrab/spec/spec_helper.rb#L1-L23 
So all credit to @mfurtak 🎁

joshdholtz commented on match/lib/match/runner.rb:
```diff
@@ -236,10 +250,16 @@ def fetch_provisioning_profile(params: nil, certificate_id: nil, app_identifier:
       uuid = parsed["UUID"]
 
+      if params[:export_path]
+        FileUtils.cp(profile, params[:export_path])
```
Do we want to do a `FileUtils.mkdir_p(params[:export_path])` here? There is a good chance a user might put a directory that doesn't exist 🙃 Either this or we need to verify on the `ConfigItem` that the path exists.

janpio replied:
Do we have any similar options for other actions where we could "steal" the behavior for consistency? I actually forgot to finish the `ConfigItem` here to be honest, because I couldn't think of where to look for how this could/should work.

janpio replied:
`cert`, `pem` and `sigh` all have an `output_path` option, so I renamed this one here as well. I also copied how those other actions create the directory at the beginning of `run` (only this time only if the param is supplied, as for the others it has a default value).

joshdholtz reviewed (changes requested):
This looks super good 🙌 Left some comments on some minor things but everything seems to check out ✅ 

I think we need a rebase in here too 🙃

janpio commented on match/lib/match/runner.rb:
```diff
@@ -202,7 +212,9 @@ def fetch_provisioning_profile(params: nil, certificate_id: nil, app_identifier:
       profiles = Dir[File.join(base_dir, "#{profile_name}.mobileprovision")]
-      keychain_path = FastlaneCore::Helper.keychain_path(params[:keychain_name]) unless params[:keychain_name].nil?
+      if Helper.mac?
+        keychain_path = FastlaneCore::Helper.keychain_path(params[:keychain_name]) unless params[:keychain_name].nil?
```
`keychain_path` does not exist on non-Mac, so it being nil is correct.
`device_count_different?` has `keychain_path` as default `nil` anyway, so this should be fine, correct?

janpio replied:
ping

janpio commented on match/lib/match/runner.rb:
```diff
@@ -220,11 +232,13 @@ def fetch_provisioning_profile(params: nil, certificate_id: nil, app_identifier:
           UI.error("A new one cannot be created because you enabled `readonly`")
-          UI.error("Provisioning profiles in your repo for type `#{prov_type}`:")
-          all_profiles.each { |p| UI.error("- '#{p}'") }
+          if Dir.exist?(base_dir)
```
> What is the reasonging why `base_dir`wouldn't exist here?

When you run `match` the first time for a specific `prov_type` (appstore, development, adhoc, enterprise) the folder referenced in `base_dir` does not exist yet, so we can't list its entries.

> Is this the fix for #13965? If so, can we get a comment here saying that 😇

Indeed, I mentioned the issue in the commit so it gets closed when this is merged.

Should I mention the issue in the comment or the reason behind `base_dir` being possible empty?

```suggestion
           if Dir.exist?(base_dir) # folder for `prov_type` does not exist on first match use for that type
```

janpio committed:
Update match/lib/match/runner.rb

janpio committed:
make sure `output_path` exists

janpio committed:
rename export_path => output_path (1)

janpio committed:
rename export_path => output_path (2)

janpio committed:
rubocop

joshdholtz reviewed (approved):
@janpio This looks good and works good! Sorry about taking so long to get this through 😔
